src/h3/mod.rs

```diff
@@ -26,4 +26,778 @@
+    }
+
+    /// Returns an available unidirectional stream ID
+    /// for the local endpoint to use
+    fn get_available_uni_stream(&mut self) -> Result<u64> {
+        if self.highest_uni_stream_id < std::u64::MAX {
+            let ret = self.highest_uni_stream_id;
+            self.highest_uni_stream_id += 4;
+            return Ok(ret);
+        }
+
+        Err(Error::LimitExceeded)
+    }
+
+    /// Open HTTP/3 control stream. Nothing happens if stream
+    /// is already open
+    fn open_control_stream(
+        &mut self, quic_conn: &mut super::Connection,
+    ) -> Result<()> {
+        if self.control_stream_id.is_none() {
+            let stream_id = self.get_available_uni_stream()?;
+            info!(
+                "{} Opening Control stream on id {}",
+                quic_conn.trace_id(),
+                stream_id
+            );
+            quic_conn.stream_send(
+                stream_id,
+                &stream::HTTP3_CONTROL_STREAM_TYPE_ID.to_be_bytes(),
+                false,
+            )?;
+
+            self.control_stream_id = Some(stream_id);
+        }
+
+        Ok(())
+    }
+
+    /// Open QPACK encoder and decoder streams. Nothing happens
+    /// if streams are already open
+    fn open_qpack_streams(
+        &mut self, quic_conn: &mut super::Connection,
+    ) -> Result<()> {
+        if self.local_qpack_streams.encoder_stream_id.is_none() {
+            let stream_id = self.get_available_uni_stream()?;
+            info!(
+                "{} Opening QPACK Encoder stream on id {}",
+                quic_conn.trace_id(),
+                stream_id
+            );
+            quic_conn.stream_send(
+                stream_id,
+                &stream::QPACK_ENCODER_STREAM_TYPE_ID.to_be_bytes(),
+                false,
+            )?;
+
+            // TODO await ACK of stream open?
+            self.local_qpack_streams.encoder_stream_id = Some(stream_id);
+        }
+
+        if self.local_qpack_streams.decoder_stream_id.is_none() {
+            let stream_id = self.get_available_uni_stream()?;
+            info!(
+                "{} Opening QPACK Decoder stream on id {}",
+                quic_conn.trace_id(),
+                stream_id
+            );
+            quic_conn.stream_send(
+                stream_id,
+                &stream::QPACK_DECODER_STREAM_TYPE_ID.to_be_bytes(),
+                false,
+            )?;
+
+            // TODO await ACK of stream open?
+            self.local_qpack_streams.decoder_stream_id = Some(stream_id);
+        }
+
+        Ok(())
+    }
+
+    /// Send SETTINGS frame based on HTTP/3 configuration
+    fn send_settings(&mut self, quic_conn: &mut super::Connection) -> Result<()> {
+        self.open_control_stream(quic_conn)?;
+
+        let mut d = [42; 128];
+
+        // Client cannot send placeholders, so validate here
+        let num_placeholders = if self.is_server {
+            self.local_settings.num_placeholders
```

Comment:
Isn't this an Option already? So you can just use it directly. It shouldn't be Some for clients.